src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,735 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
-		uint name_left  = rtl ? left : left + type_icon_width;
-		uint name_right = rtl ? right - type_icon_width : right;
+		uint count_start = this->vscroll->GetPosition();
+		uint count_end = count_start + this->vscroll->GetCapacity();
 
-		int i = 0;
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			TextColour colour;
-			if (this->selected_item == i++) { // Selected item, highlight it
-				GfxFillRect(r.left + 1, y, r.right - 1, y + this->line_height - 1, PC_BLACK);
-				colour = TC_WHITE;
-			} else {
-				colour = TC_BLACK;
+		uint y = top + (this->line_height * (count - count_start));
+
+		/* Draw the company line (if in range of scrollbar). */
+		if (count >= count_start && count < count_end) {
+			uint x = rtl ? text_left : text_right;
+
+			/* If there are buttons for this company, draw them. */
+			auto button_find = this->buttons.find(count);
+			if (button_find != this->buttons.end()) {
+				x = this->DrawButtons(x, y, &button_find->second);
 			}
 
-			if (ci->client_id == CLIENT_ID_SERVER) {
-				DrawString(type_left, type_right, y + text_offset, STR_NETWORK_SERVER, colour);
+			if (company_id == COMPANY_SPECTATOR) {
+				DrawSprite(SPR_COMPANY_ICON, PALETTE_TO_GREY, rtl ? right - d.width - 4 : left + 4, y + offset);
+				DrawString(rtl ? x : text_left, rtl ? text_right : x, y + text_y_offset, STR_NETWORK_CLIENT_LIST_SPECTATORS, TC_SILVER);
+			} else if (company_id == COMPANY_NEW_COMPANY) {
+				DrawSprite(SPR_COMPANY_ICON, PALETTE_TO_GREY, rtl ? right - d.width - 4 : left + 4, y + offset);
+				DrawString(rtl ? x : text_left, rtl ? text_right : x, y + text_y_offset, STR_NETWORK_CLIENT_LIST_NEW_COMPANY, TC_WHITE);
 			} else {
-				DrawString(type_left, type_right, y + text_offset, STR_NETWORK_CLIENT, colour);
+				DrawCompanyIcon(company_id, rtl ? right - d.width - 4 : left + 4, y + offset);
+
+				SetDParam(0, company_id);
+				SetDParam(1, company_id);
+				DrawString(rtl ? x : text_left, rtl ? text_right : x, y + text_y_offset, STR_COMPANY_NAME, TC_SILVER);
 			}
+		}
 
-			/* Filter out spectators */
-			if (Company::IsValidID(ci->client_playas)) DrawCompanyIcon(ci->client_playas, icon_left, y + icon_offset);
+		y += this->line_height;
+		count++;
+
+		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
+			if (ci->client_playas != company_id) continue;
+			if (StrEmpty(ci->client_name)) continue;
 
-			DrawString(name_left, name_right, y + text_offset, ci->client_name, colour);
+			/* Draw the player line (if in range of scrollbar). */
+			if (count >= count_start && count < count_end) {
+				uint x = rtl ? text_left : text_right;
 
-			y += line_height;
+				/* If there are buttons for this client, draw them. */
+				auto button_find = this->buttons.find(count);
+				if (button_find != this->buttons.end()) {
+					x = this->DrawButtons(x, y, &button_find->second);
+				}
+
+				if (ci->client_id == _network_own_client_id) {
+					Dimension d2 = GetSpriteSize(SPR_PLAYER_SELF);
```

Comment:
Maybe put this in a `DrawPlayerIcon(bool rtl, int text_right, int text_left, int y, SpriteID sprite)`. That way the code does not get duplicated and any future updates to alignment should be easier.
Alternatively determine the sprite based on the client id, e.g. `SpriteID GetPlayerIcon(ClientID client_id)` or just the `if`-statement in the code, and then depending on whether the SpriteID is valid or not perform the logic of the branch with that sprite.